homeassistant/components/huawei_lte/binary_sensor.py

```diff
@@ -0,0 +1,128 @@
+        entities.append(HuaweiLteMobileConnectionBinarySensor(router))
+
+    async_add_entities(entities, True)
+
+
+@attr.s
+class HuaweiLteBaseBinarySensor(HuaweiLteBaseEntity, BinarySensorDevice):
+    """Huawei LTE binary sensor device base class."""
+
+    key: str
+    item: str
+    _raw_state: Optional[str] = attr.ib(init=False, default=None)
+
+    async def async_added_to_hass(self):
+        """Subscribe to needed data on add."""
+        await super().async_added_to_hass()
+        self.router.subscriptions[self.key].add(f"{BINARY_SENSOR_DOMAIN}/{self.item}")
+
+    async def async_will_remove_from_hass(self):
+        """Unsubscribe from needed data on remove."""
+        await super().async_will_remove_from_hass()
+        self.router.subscriptions[self.key].remove(
+            f"{BINARY_SENSOR_DOMAIN}/{self.item}"
+        )
+
+    async def async_update(self):
+        """Update state."""
+        try:
+            value = self.router.data[self.key][self.item]
+        except KeyError:
+            _LOGGER.debug("%s[%s] not in data", self.key, self.item)
+            self._available = False
+            return
+        self._available = True
+        self._raw_state = str(value)
+
+
+CONNECTION_STATE_ATTRIBUTES = {
+    str(ConnectionStatusEnum.CONNECTING): "Connecting",
+    str(ConnectionStatusEnum.DISCONNECTING): "Disconnecting",
+    str(ConnectionStatusEnum.CONNECT_FAILED): "Connect failed",
+    str(ConnectionStatusEnum.CONNECT_STATUS_NULL): "Status not available",
+    str(ConnectionStatusEnum.CONNECT_STATUS_ERROR): "Status error",
+}
+
+
+@attr.s
+class HuaweiLteMobileConnectionBinarySensor(HuaweiLteBaseBinarySensor):
+    """Huawei LTE mobile connection binary sensor."""
+
+    def __attrs_post_init__(self):
+        """Initialize identifiers."""
+        self.key = KEY_MONITORING_STATUS
+        self.item = "ConnectionStatus"
+
+    @property
+    def device_class(self):
+        """Return device class."""
+        return DEVICE_CLASS_CONNECTIVITY
+
+    @property
+    def _entity_name(self) -> str:
+        return "Mobile connection"
+
+    @property
+    def _device_unique_id(self) -> str:
+        return f"{self.key}.{self.item}"
+
+    @property
+    def is_on(self) -> bool:
+        """Return whether the binary sensor is on."""
+        return self._raw_state and int(self._raw_state) in (
+            ConnectionStatusEnum.CONNECTED,
+            ConnectionStatusEnum.DISCONNECTING,
+        )
+
+    @property
+    def assumed_state(self) -> bool:
+        """Return True if real state is assumed, not known."""
+        return not self._raw_state or int(self._raw_state) not in (
+            ConnectionStatusEnum.CONNECT_FAILED,
+            ConnectionStatusEnum.CONNECTED,
+            ConnectionStatusEnum.DISCONNECTED,
+        )
+
+    @property
+    def icon(self):
+        """Return binary sensor icon."""
+        return "mdi:signal" if self.is_on else "mdi:signal-off"
```

Comment:
Device class entity icons should not be overridden.

Reply:
This is already being done for pretty much all Huawei LTE sensors and switches.

Reply:
To elaborate:

If I leave this out, the default icon for a connectivity class binary sensor is something clearly related to wired connections, which is bad for a mobile connectivity sensor.

It would be also very different from the closely related mobile data switch, which would be bad too. Now it's the same, which is consistent and good.

If I remove it from mobile data switch as well, then its icon becomes the lightning one which isn't a good one for a mobile connectivity switch, and it would also be very different from any choices I have for the mobile connection sensor, so things would get even worse.

Note also that this icon is solely for the mobile connection sensor, it's not in the base binary sensor class. The method docstring could be improved to say it's the icon for mobile connection sensor, not binary sensor.

Reply:
I'll ask for a second opinion.

Reply:
Any progress, can we move forward here?

Reply:
If we remove it, we can move forward. Otherwise we need a second opinion.

Reply:
Do we need a new device class for mobile connectivity ?

Reply:
Why not, if there's need to differentiate mobile from other connectivity for some other reason besides the icon (I don't know). But just for the icon it seems overkill to me when one can just override/"fix" the icon when need be.

Reply:
Do we have some conclusion here?

Reply:
Device classes are not just icons, they also define what data is represented. It means it impacts how systems that built on top of Home Assistant represent it, like Google, Alexa or Almond. 

So if the icon does not represent the device class properly, we should consider updating the device class icon

Reply:
For generic connectivity, I don't think there's anything wrong per se with the current icon. But for mobile connectivity, it's not a good one. And I don't want to get into the bikeshed whether one better describing mobile connectivity (e.g. the one I used here) would be acceptable for all connectivity.

I guess a kind of a sweet spot would be to be able to define a mobile connectivity device class that would inherit from the generic connectivity device class and have a different icon (possibly also some other differences/additional behavior or data sometime), although I suppose (without digging around) that that's not an option with the current device class implementation. And I think it's out of the scope for this pull request, and neither would I want to wait for it to come up before getting this in.

So at the moment my practical choices boil down to:
1. use connectivity device class, override icon (this is what the PR does now)
2. drop connectivity device class altogether, keep the icon I like
3. use connectivity device class, remove icon override (this is what Martin suggested)
4. wait for a new device class with the icon I like to be introduced

My favourite, by far, would be 1, then 4 sometime later when/if it's available. Going through 2 or 3 eventually to 4 when it's around would in my opinion in the meantime introduce either suboptimal user experience with the icon as elaborated earlier, or possible loss of some functionality caused by a missing device class. Just waiting for 4 before introducing this is an unnecessary delay.

If this can't go in its current form (i.e. 1 above) I'm currently torn between removing the icon and getting this approved, or just keeping the implementation to myself until 4 comes up. But I'd appreciate a clear decision/conclusion.

Reply:
I prefer to go with 2 while we wait till 4 is available. The device classes are not extensively used. I think the only thing that I know that uses it now is Almond, for which you'll soon be able to ask if X is connected. Changing a device class is technically a breaking change, adding one is not. 

We don't differentiate or inherit device classes, because the extra complexity will cause pain in other systems relying on it.

Reply:
Device class dropped.

At least in one sense inheriting a device class will cause _less_ complexity to consumers than adding a new one. For example if "mobile connectivity" was extended from "connectivity", systems interesting in any kind of connectivity would just have to check for "connectivity" instead of both.